File: frontend/src/types/refactor.ts
export interface BreakingChange {
  entity: string
  impact: 'HIGH' | 'MEDIUM' | 'LOW'
  reason: string
  affectedCallers: string[]
}

export interface BreakingChangesResult {
  detected: boolean
  changes: BreakingChange[]
}

export interface RefactorProposal {
  proposalId: string
  entityId: string
  entityName: string
  filePath: string
  original: string
  proposed: string
  rationale: string[]
  behavioralDifferences: string[]
  breakingChanges?: BreakingChangesResult
  originalChecksum: string
}

export interface RefactorProposalEnvelope {
  data: RefactorProposal | null
  error: { code: string; message: string } | null
}

File: backend/app/schemas/refactor.py
"""Pydantic schemas for refactor proposal API (T-17)."""

import uuid

from pydantic import BaseModel, ConfigDict, Field, alias_generators

from app.schemas.common import Envelope


class BreakingChange(BaseModel):
    model_config = ConfigDict(
        alias_generator=alias_generators.to_camel, populate_by_name=True
    )

    entity: str
    impact: str  # HIGH, MEDIUM, LOW
    reason: str
    affected_callers: list[str] = Field(default_factory=list)


class BreakingChangesResult(BaseModel):
    model_config = ConfigDict(
        alias_generator=alias_generators.to_camel, populate_by_name=True
    )

    detected: bool
    changes: list[BreakingChange] = Field(default_factory=list)


class RefactorProposal(BaseModel):
    model_config = ConfigDict(
        alias_generator=alias_generators.to_camel, populate_by_name=True
    )

    proposal_id: uuid.UUID
    entity_id: uuid.UUID
    entity_name: str
    file_path: str
    original: str
    proposed: str
    rationale: list[str] = Field(default_factory=list)
    behavioral_differences: list[str] = Field(default_factory=list)
    breaking_changes: BreakingChangesResult = Field(
        default_factory=lambda: BreakingChangesResult(detected=False, changes=[])
    )
    # SHA-256 checksum of the original source — proves original repo unchanged
    original_checksum: str


RefactorProposalEnvelope = Envelope[RefactorProposal]

